Fixed #26293 - Moved append_slash logic outside of prepend_www logic

ieatkittens opened the pull request:
Moving logic required dynamically building the redirect_url as well as updating test_non_ascii_query_string_does_not_crash to expect a 301 from process_request instead of expecting no redirect.

Also added a regression test (and renamed another test to differentiate the two).

timgraham commented on tests/middleware/tests.py:
```diff
@@ -301,7 +314,7 @@ def test_non_ascii_query_string_does_not_crash(self):
         request.META['QUERY_STRING'] = force_str('drink=café')
         r = CommonMiddleware().process_request(request)
-        self.assertIsNone(r)
+        self.assertEqual(r.status_code, 301)
```
Since `process_request()` returns a response, the second part of the test isn't needed (see 434d309ef6dbecbfd2b322d3a1da78aa5cb05fa8)

timgraham commented on tests/middleware/tests.py:
```diff
@@ -62,9 +62,22 @@ def test_append_slash_slashless_unknown(self):
+        self.assertEqual(r.url, '/slash/')
+
+    @override_settings(APPEND_SLASH=True)
+    def test_append_slash_redirect_response(self):
```
As above, since `process_request()` returns a response, the second test isn't needed (see 434d309ef6dbecbfd2b322d3a1da78aa5cb05fa8).

timgraham commented:
Some edits: http://dpaste.com/2J098KX
We should probably backport to 1.9 since it's a regression. Could you add a release note for 1.9.5 and try to rephrase the commit message to describe the issue that's fixed rather than the implementation? Thanks!

ieatkittens commented:
I've updated the tests as you requests, changed the commit message and added release notes under 1.9.5 - let me know if anything else needs to be updated/changed!

ieatkittens commented:
Oh, I totally missed your edits.  Will add those later.  Thanks for your feedback!

timgraham commented:
merged in 9390da7fb6e251eaa9a785692f987296cb14523f, thanks!
